src/ZkFold/Symbolic/Data/ByteString.hs

```diff
@@ -309,57 +271,114 @@ instance
+  , Arithmetic a
+  ) => Concat (ByteString m (ArithmeticCircuit a)) (ByteString k (ArithmeticCircuit a)) where
+
+    concat bs = ByteString $ bsCircuit {acOutput = bsOutputs}
+        where
+            bsCircuit = Haskell.mconcat $ (\(ByteString bits) -> bits {acOutput = U1}) <$> bs
+
+            bsOutputs :: Vector k Natural
+            bsOutputs = V.unsafeConcat @(Div k m) $ (\(ByteString bits) -> acOutput bits) <$> bs
+
+instance
+  ( KnownNat n
+  , n <= m
+  ) => Truncate (ByteString m (ArithmeticCircuit a)) (ByteString n (ArithmeticCircuit a)) where
+
+    truncate (ByteString bits) = ByteString $ bits { acOutput = V.take @n (acOutput bits) }
+
 
 -- | Only a bigger ByteString can be truncated into a smaller one.
 --
 instance
-  ( KnownNat m
+  ( Symbolic (Interpreter (Zp p))
+  , KnownNat m
   , KnownNat n
   , n <= m
-  , Finite (Zp p)
   ) => Truncate (ByteString m (Interpreter (Zp p))) (ByteString n (Interpreter (Zp p))) where
 
     truncate = fromConstant @Natural . (`shiftR` diff) . toConstant
         where
             diff :: Haskell.Int
             diff = Haskell.fromIntegral $ getNatural @m Haskell.- getNatural @n
 
--- | Only a smaller ByteString can be extended into a bigger one.
---
-instance
-  ( KnownNat n
-  , m <= n
-  , Finite (Zp p)
-  ) => Extend (ByteString m (Interpreter (Zp p))) (ByteString n (Interpreter (Zp p))) where
+--------------------------------------------------------------------------------
+instance (Symbolic (Interpreter (Zp p)), c ~ Interpreter (Zp p), KnownNat n) => ShiftBits (ByteString n (Interpreter (Zp p))) where
+    shiftBits bs@(ByteString oldBits) s
+      | s == 0 = bs
+      | Haskell.abs s >= Haskell.fromIntegral (getNatural @n) = false
+      | otherwise = ByteString $ symbolicF oldBits (\v ->  V.shift v s (fromConstant (0 :: Integer))) solve
+      where
+        solve :: forall i m. MonadCircuit i (BaseField c) m => Vector n i -> m (Vector n i)
+        solve bitsV = do
+            let bits = V.fromVector bitsV
+            zeros <- replicateM (Haskell.fromIntegral $ Haskell.abs s) $ newAssigned (Haskell.const zero)
 
-    extend = fromConstant @Natural . toConstant
+            let newBits = case s < 0 of
+                        Haskell.True  -> take (Haskell.fromIntegral $ getNatural @n) $ zeros <> bits
+                        Haskell.False -> drop (Haskell.fromIntegral s) $ bits <> zeros
 
-instance Finite (Zp p) => BitState ByteString n (Interpreter (Zp p)) where
-    isSet (ByteString (Interpreter v)) ix = Bool (Interpreter . Par1 . (!! ix) . V.fromVector $ v)
-    isUnset bs ix = let Bool (Interpreter zp) = isSet bs ix
-                     in Bool (Interpreter $ (one -) <$> zp)
+            pure $ V.unsafeToVector newBits
 
---------------------------------------------------------------------------------
 
-instance (Arithmetic a, KnownNat n) => ShiftBits (ByteString n (ArithmeticCircuit a)) where
+    -- | rotateBits does not even require operations on the circuit.
```

Comment:
If it does not require operations on the circuit, this can be done via functions from `HApplicative` & `Package`